src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ValidationAttributeProvider.cs

```diff
@@ -0,0 +1,111 @@
+        private readonly IClientModelValidatorProvider _clientModelValidatorProvider;
+
+        /// <summary>
+        /// Initializes a new <see cref="ValidationAttributeProvider"/> instance.
+        /// </summary>
+        /// <param name="optionsAccessor">The accessor for <see cref="MvcViewOptions"/>.</param>
+        /// <param name="metadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        /// <param name="clientValidatorCache">The <see cref="ClientValidatorCache"/> that provides
+        /// a list of <see cref="IClientModelValidator"/>s.</param>
+        public ValidationAttributeProvider(
+            IOptions<MvcViewOptions> optionsAccessor,
+            IModelMetadataProvider metadataProvider,
+            ClientValidatorCache clientValidatorCache)
+        {
+            if (optionsAccessor == null)
+            {
+                throw new ArgumentNullException(nameof(optionsAccessor));
+            }
+
+            if (metadataProvider == null)
+            {
+                throw new ArgumentNullException(nameof(metadataProvider));
+            }
+
+            if (clientValidatorCache == null)
+            {
+                throw new ArgumentNullException(nameof(clientValidatorCache));
+            }
+
+            _clientValidatorCache = clientValidatorCache;
+            _metadataProvider = metadataProvider;
+
+            var clientValidatorProviders = optionsAccessor.Value.ClientModelValidatorProviders;
+            _clientModelValidatorProvider = new CompositeClientModelValidatorProvider(clientValidatorProviders);
+        }
+
+        /// <inheritdoc />
+        public virtual void AddValidationAttributes(
+            ViewContext viewContext,
+            ModelExplorer modelExplorer,
+            string expression,
+            IDictionary<string, string> attributes)
+        {
+            if (viewContext == null)
+            {
+                throw new ArgumentNullException(nameof(viewContext));
+            }
+
+            if (attributes == null)
+            {
+                throw new ArgumentNullException(nameof(attributes));
+            }
+
+            // Only render attributes if client-side validation is enabled, and then only if we've
+            // never rendered validation for a field with this name in this form.
+            var formContext = viewContext.ClientValidationEnabled ? viewContext.FormContext : null;
+            if (formContext == null)
+            {
+                return;
+            }
+
+            var fullName = viewContext.ViewData.TemplateInfo.GetFullHtmlFieldName(expression);
+            if (formContext.RenderedField(fullName))
+            {
+                return;
+            }
+
+            formContext.RenderedField(fullName, true);
+
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
```

Comment:
Are you suggesting this method should do less than answer the question "what validation attributes would a HTML or tag helper add for expression X?" Because the code above covers the cases where the answer to that question is "none".

But, apart from the inefficiency of creating a `ModelExplorer` for expressions when e.g. client-side validation is disabled, I could move this particular line into the `DefaultHtmlGenerator` code that calls this.

Further thoughts?